File: doc/example_6.c
/* This example uses the dp_zsurface function to produce a height field
 * or z surface from a 2D array of data.  The dp_axis and dp_boundbox
 * routines are used to provide an axis and bounds for the surface.
 */
#include <stdio.h>
#include <math.h>
#include <p3dgen.h>
#include <drawp3d.h>

/* Macro to ease checking of error return codes */
#define ERRCHK( string ) if (!(string)) fprintf(stderr,"ERROR!\n")

/* Information giving the corners of the bounding box and ends of the axis */
static P_Point corner1 = { -2.0, -2.0, -1.5 };
static P_Point corner2 = { 2.0, 2.0, 1.5 };
static P_Point axis_pt1 = { -2.0, -2.1, -2.0 };
static P_Point axis_pt2 = { 2.0, -2.1, -2.0 };
static P_Vector axis_up = { 0.0, 0.0, 1.0 };

/* Camera information */
static P_Point lookfrom= { 0.0, -10.0, 2.0 };
static P_Point lookat= { 0.0, 0.0, 0.0 };
static P_Vector up= { 0.0, 0.0, 1.0 };
static float fovea= 45.0;
static float hither= -5.0;
static float yon= -20.0;

/* Colors to be used for the bounding box and axis */
static P_Color red_color= { P3D_RGB, 1.0, 0.0, 0.0, 1.0 };
static P_Color blue_color= { P3D_RGB, 0.0, 0.0, 1.0, 1.0 };

/* This function provides data to use to produce the z surface. */
static float *create_zsurfdat( z )
float z[29][31];
{
  float r, x, y, tz;
  int i, j;

  for (i=0; i<29; i++) {
    x= (float)(i-14);
    for (j=0; j<31; j++) {
      y= (float)(j-15);
      r= sqrt( x*x + y*y );
      tz= cos(r);
      z[i][j]= 2.0*exp(-r/5.0)*tz;
    }
  }
}

/* Those data values for which *exclude is set to 1 will be excluded
 * from the z surface, leaving holes.
 */
static void testfun( int *exclude, float *testval, int *x, int *y )
{
  if (*testval < 1.9 )
    *exclude = 0;
  else 
    *exclude = 1;
}

/* This routine actually creates the height field or z surface. */
static void create_surf()
{
  float z[29][31];

  create_zsurfdat(z);

  /* This sets the color map which will be used to map the value data
   * to colors on the z surface.  We will be using the surface values
   * themselves, which give the height of the surface at any given
   * point, to color the surface.
   */
  ERRCHK( dp_std_cmap( -2.0, 2.0, 2 ) );

  /* Open a GOB to contain the surface, bounding box, and axis */
  ERRCHK( dp_open("zsurf_gob") );

  /* Insert a bounding box, with an associated color. */
  ERRCHK( dp_open("") );
  ERRCHK( dp_gobcolor( &blue_color ) );
  ERRCHK( dp_boundbox( &corner1, &corner2 ) );
  ERRCHK( dp_close() );

  /* Insert an axis, with an associated color.  See the reference manual
   * for details on the calling parameters.  
   */
  ERRCHK( dp_open("") );
  ERRCHK( dp_gobcolor( &red_color ) );
  ERRCHK( dp_axis( &axis_pt1, &axis_pt2, &axis_up, -2.0, 2.0, 3, 
                   "Sample Z surface", 0.25, 2 ) );
  ERRCHK( dp_close() );

  /* Create the z surface, inserting it into the GOB "zsurf_gob".  The
   * values in z are used both to provide the heights for the surface
   * and to provide values with which to color it.  testfun provides
   * a function which can be used to exclude points from the surface;
   * it could be left null if all points were to be included.
   */
  ERRCHK( dp_zsurface(P3D_CVVTX, (float *)z, (float *)z, 29, 31, 
		      &corner1, &corner2, testfun ) );

  /* Close the GOB "zsurf_gob". */
  ERRCHK( dp_close() );
}

main()
/* This is the main routine. */
{
  /* Initialize the renderer. */
  ERRCHK( dp_init_ren("myrenderer","p3d","example_6.p3d",
		      "I don't care about this string") );

  /* Create a camera, according to specs given in definitions above. */
  ERRCHK( dp_camera("mycamera",&lookfrom,&lookat,&up,fovea,hither,yon) );

  /* Create a GOB called "zsurf_gob", containing a z surface, a bounding
   * box, and an axis.
   */
  create_surf();

  /* Cause the GOB to be rendered. */
  ERRCHK( dp_snap( "zsurf_gob", "standard_lights", "mycamera" ) );

  /* Shut down the DrawP3D software. */
  ERRCHK( dp_shutdown() );
}

File: doc/example_6.f
C Example 6:
C This example uses the PZSURF function to produce a height field
C or Z surface from a 2D array of data.  The PAXIS and PBNDBX 
C routines are used to provide an axis and bounds for the surface.

C Subroutine to ease checking of return codes.
      subroutine errchk( ival )
      integer ival
      if (ival.ne.1) write(6,*) 'ERROR!'
      return
      end

      subroutine mkzsur( zdat )
C
C     This routine prepares data used to define the surface.
C
      real zdat( 29, 31 )
      integer i, j
      real r, tz
      real ir,jr
      do 10, i= -14, 14
         do 20, j = -15, 15
            ir = i
            jr = j
            r = sqrt( ir*ir + jr*jr )
            tz = cos(r)
            zdat(i+15,j+16) = 2.0*exp(-r/5.0)*tz
 20      continue
 10   continue
      return
      end
      

      subroutine tfun( bool, val, x, y )
C
C     Those data values for which bool is set to 1 will be excluded
C     from the z surface, leaving holes.
C
      integer bool
      real val
      integer x, y
      if (val .lt. 1.9) then 
         bool=0
      else
         bool=1
      endif
      return
      end

      subroutine zsfex()
C
C     This routine actually creates the height field or z surface.
C
      integer popen, pclose, pgbclr, paxis, pbndbx, psdcmp, pzsurf
      external tfun
      real zdat(29,31)

C     Definitions for some needed colors.
      real red(4)
      data red/1.0,0.0,0.0,1.0/
      real blue(4)
      data blue/0.0,0.0,1.0,1.0/

C     Specification of the corners of the bounding box and z surface
      real crna(3)
      data crna/-2.0, -2.0, -1.5/
      real crnb(3)
      data crnb/2.0, 2.0, 1.5/

C     This sets the location and orientation of the axis.
      real axspta(3)
      data axspta/-2.0,-2.1,-2.0 /
      real axsptb(3)
      data axsptb/2.0,-2.1,-2.0 /
      real axisup(3)
      data axisup/0.0,0.0,1.0 /

C     Calculate the data defining the surface.
      call mkzsur(zdat) 

C     This sets the color map which will be used to map the value data
C     to colors on the z surface.  We will be using the surface values
C     themselves, which give the height of the surface at any given
C     point, to color the surface.
      call errchk( psdcmp(-2.0,2.0,2) )

C     Open a GOB to contain the surface, bounding box, and axis.
      call errchk( popen('zsurf_gob') )

C     Insert a bounding box, with an associated color.
      call errchk( popen(' ') )
      call errchk( pbndbx( crna, crnb ) )
      call errchk( pgbclr(0,blue(1), blue(2), blue(3), blue(4) ) )
      call errchk( pclose() )

C     Insert an axis, with an associated color.  See the reference
C     manual for details on the calling parameters.
      call errchk( popen(' ') )
      call errchk( paxis( axspta, axsptb, axisup, -2.0, 2.0, 3,
     $      'BoundingBox',0.25, 2 ) ) 
      call errchk( pgbclr(0,red(1), red(2), red(3), red(4) ) )
      call errchk( pclose() )

C     Create the z surface, inserting it into the GOB 'zsurf_gob'.
C     The values in zdat are used both to provide the heights for
C     the surface and to provide values with which to color it.  
C     tfun provides a function which can be used to exclude points
C     from the surface;  it could be set to 0 and ignored if the
C     previous parameter were 1.
      call errchk( pzsurf( 4, zdat, zdat, 29, 31, crna, crnb, 
     $     0, tfun ) )

C     Close the GOB 'zsurf_gob'.
      call errchk( pclose() )

      return
      end

      program ex6
C
C     This is the main program
C
      integer pintrn, pcamra, psnap, pshtdn

C     The following information sets the camera position.
      real lookfm(3), lookat(3), up(3), fovea, hither, yon
      data lookfm/ 0.0, -10.0, 2.0 /
      data lookat/ 0.0, 0.0, 0.0 /
      data up/ 0.0, 0.0, 1.0 /
      data fovea, hither, yon/45.0, -5.0, -20.0/

C     Initialize the renderer.
      call errchk( pintrn('myrenderer','p3d','example_6.p3d',
     $     'c') )

C     Create a camera, according to the specs given above.
      call errchk( pcamra('mycamera',lookfm,lookat,up,fovea,
     $     hither,yon) )

C     Create a z surface, bounding box, and axis in a GOB called
C     'zsurf_gob'.
      call zsfex()

C     Cause the GOB to be rendered.
      call errchk( psnap('zsurf_gob','standard_lights', 'mycamera'))

C     Shut down the DrawP3D software.
      call errchk( pshtdn() )

      stop
      end

